x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -83,47 +100,120 @@
+
+        // TODO: We should have a way to express true Zero applications
+        final Set<String> applicationNamesCacheKey = (isEmpty(applications) || applications.contains("*")) ?
+            Collections.singleton("*") : Set.copyOf(applications);
+
+        // Always fetch for the concrete application names even when the passed in application names do not
+        // contain any wildcard. This serves as a negative lookup.
+        Set<String> concreteApplicationNames = applicationNamesCache.get(applicationNamesCacheKey);
+
+        if (concreteApplicationNames != null && concreteApplicationNames.size() == 0) {
+            logger.debug("returning empty application privileges as application names result in empty list");
+            listener.onResponse(Collections.emptySet());
+
+        } else {
+            final Tuple<Set<String>, Map<String, Set<ApplicationPrivilegeDescriptor>>> cacheStatus;
+            if (concreteApplicationNames == null) {
+                cacheStatus = cacheStatusForApplicationNames(applicationNamesCacheKey);
+            } else {
+                cacheStatus = cacheStatusForApplicationNames(concreteApplicationNames);
+            }
+
+            if (cacheStatus.v1().isEmpty()) {
+                logger.debug("All application privileges found in cache");
+                final Set<ApplicationPrivilegeDescriptor> cachedDescriptors =
+                    cacheStatus.v2().values().stream().flatMap(Collection::stream).collect(Collectors.toUnmodifiableSet());
+                listener.onResponse(filterDescriptorsForNames(cachedDescriptors, names));
+            } else {
+                final long invalidationCounter = numInvalidation.get();
+                // Always fetch all privileges of an application for caching purpose
+                logger.debug("Fetching application privilege documents for: {}", cacheStatus.v1());
+                innerGetPrivileges(cacheStatus.v1(), ActionListener.wrap(fetchedDescriptors -> {
+                    final Map<String, Set<ApplicationPrivilegeDescriptor>> mapOfFetchedDescriptors = fetchedDescriptors.stream()
+                        .collect(Collectors.groupingBy(ApplicationPrivilegeDescriptor::getApplication, Collectors.toUnmodifiableSet()));
+                    final Set<String> allApplicationNames =
+                        Stream.concat(cacheStatus.v2().keySet().stream(), mapOfFetchedDescriptors.keySet().stream())
+                            .collect(Collectors.toUnmodifiableSet());
+                    // Avoid caching potential stale results.
+                    // TODO: It is still possible that cache gets invalidated immediately after the if check
+                    if (invalidationCounter == numInvalidation.get()) {
+                        // Do not cache the names if expansion has no effect
+                        if (allApplicationNames.equals(applicationNamesCacheKey) == false) {
```

Comment:
This is a minor optimisation so we do not store an etnry of `{Set('app-1'), Set('app-1')}`, where there is no expansion at all.